### PR TITLE
Fix dataloader overhead when restarting from a checkpoint

### PR DESCRIPTION
**Pull request by matthewdcong**
Currently when we restart from a checkpoint we iterate over the `trainloader` and skip the body of the loop if the step is less than the step in the checkpoint. This continues until we get to the step indicated in the checkpoint.

However, this is **extremely slow** because `DataLoader.__next__` _loads the data from disk_ which means we're incurring an expensive overhead for disk I/O and then just discarding the result of the load for any epochs/batches we've already trained. Instead, we should iterate over the `trainloader.sampler` which merely samples the index without actually performing the load while preserving the same RNG semantics as iterating over `trainloader` itself. On a large dataset which has already been trained for a number of epochs, this can save many hours.

Furthermore, restarting from a checkpoint is missing the RNG seed initialization resulting in non-deterministic minibatch ordering upon restart. This PR addresses that as well.

### CODE REVIEW

**Review by harrism — APPROVED** [review]
Nice.
